[Builder] Introduce a typed command system and 2 phase parse/dispatch build

simonferquel opened the pull request:
This is a work base to introduce more features like build time
dockerfile optimisations, dependency analysis and parallel build, as
well as a first step to go from a dispatch-inline process to a
frontend+backend process.

**- What I did**
I moved the whole parsing of Dockerfiles into a separate package generating a typed command graph.
Then I changed the current dispatch logic so that it processes typed commands instead of AST subtrees.

This allows to have a global view of the build process before actually running it, allowing command tree optimisations, parallel execution of build stages, and even distributed execution in the future.

It has also the benefit to make the Builder code much more maintainable with a clear separation of concerns between, parsing, env variables / build args expansion and command dispatching.

**- How I did it**

I refactored the code to introduce a collection of typed commands corresponding to each operation in the dispatchers.go file.

**- How to verify it**

Builder keeps the same public interface. The current test suite is OK.
~~I still have to re-implement existing in-package unit tests though~~ <- unit tests are back

**- Description for the changelog**

Refactoring of dockerfile builder to enable future improvements

cc @tonistiigi and @dnephin 


(Note: I squashed all my commits due to rebase complexity. Unsquashed - and outdated - branch is here: https://github.com/simonferquel/docker/tree/typed-builder-commands)

simonferquel committed:
Introduce a typed command system and 2 phase parse/dispatch build

This is a work base to introduce more features like build time
dockerfile optimisations, dependency analysis and parallel build, as
well as a first step to go from a dispatch-inline process to a
frontend+backend process.

Signed-off-by: Simon Ferquel <simon.ferquel@docker.com>

tonistiigi commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,297 @@
+		return err
+	}
+	c.Dest = dst
+	from, err := expander(c.From)
```
I don't think this is supported atm. We have to be careful about this as it may prevent us from determining dependencies between stages before the image config has been pulled. We may want to do this in the context of global args like the strings in `FROM` command but that's a discussion for later.

simonferquel replied:
hmm, flags where not expanded then ? i'll change that!

tonistiigi commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,617 @@
+		switch c := cmd.(type) {
+		case *FromCommand:
+			if c.StageName == "" {
+				c.StageName = strconv.Itoa(len(stages))
```
This does not look correct. `--from=n` syntax is for accessing all stages, not the ones that didn't have a name.

tonistiigi commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,617 @@
+			if c.StageName == "" {
+				c.StageName = strconv.Itoa(len(stages))
+			}
+			stage := BuildableStage{Name: c.StageName, Commands: []interface{}{c}}
```
I'd really like to remove a from command inside a stage and just make `FROM==buildstage`. If complicates things atm it could be a follow-up with a comment.

simonferquel replied:
It complicates things a bit for the BuildFromConfig stuff: in one case we have a from command that will initialize the dispatchState with a named stage or image lookup, in the other case, the resume build command just copy an existing runConfig. Also, I like the idea of having every command dispatch logic mapped to a given typed command and for consistency sake, I like the idea to treat "FROM" command as any other dispatchable command (even if on parsing it has a slightly different behavior as it is a marker for a stage beginning)

tonistiigi replied:
`FROM` is not allowed in `BuildFromConfig`. Only commands that change image config are allowed there. Basically, it is not building anything, it is just converting parameters in dockerfile syntax to a json config syntax.

tonistiigi replied:
https://github.com/moby/moby/blob/274cc09247696c56017aed6bafb1089bd2ebbfde/builder/dockerfile/builder.go#L25-L36

simonferquel replied:
Yes, I understand, my point is that a stageBuilder can be initialized by either a `FROM` command or a `ResumeBuild` command created when doing a `BuildFromConfig`. I like to leverage the typed-command style dispatching to change this alternative behavior.
Some other things to consider:
- From is not only about initializing the stage: At dispatch time, it might trigger `onbuild` expression dispatching
- From supports buildargs expansion and cannot be entirely resolved at parse time
- From can reference another stage, in this case it is not resolvable at parse time either, as it needs to get the runconfig produced by the referenced stage dispatching
- I'd like to keep the degree of consistency I have now: 1 dockerfile instruction == 1 typed command (even meta args are typed command, they just don't belong to a specific stage)
- In term of consistency, I really like the way it is, as the builder don't need to know anything about stage initialization, it only execute one list of commands per stage. The 1st command accidently initialize the stage runConfig, but the builder itself does not even care.

tonistiigi commented:
cc @AkihiroSuda as well as this is related to https://github.com/moby/moby/issues/32550#issuecomment-297867334

AkihiroSuda commented on builder/dockerfile/builder.go:
```diff
@@ -168,62 +173,69 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
+			select {
+			case <-b.clientCtx.Done():
+				logrus.Debug("Builder: build cancelled!")
+				fmt.Fprint(b.Stdout, "Build cancelled")
```
`\n` missing?

AkihiroSuda commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,297 @@
+
+type BuildableStage struct {
+	Name     string
+	Commands []interface{}
```
`interface{}` -> `BaseCommand`?

simonferquel replied:
There is one specific command that does not embed CommandBase as it is not parsed from source code (ResumeBuild, used in BuildFromConfig). This also opens up some cool features like in the future, being able to split typed commands into more fine grained subcommands to increase parallelism opportunities. (for exemple, having the AddCommand split into many `fetch` subcommands -1 per source- and join them in a `mergeInto` command)

AkihiroSuda commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,617 @@
+
+	labels := KeyValuePairs{}
+
+	for j := 0; j < len(req.args); j++ {
```
Can we fold two `j++` into a single `j+=2`? (as in `parseEnv`)

AkihiroSuda commented on builder/dockerfile/builder.go:
```diff
@@ -168,62 +173,69 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
-		if n.Value == command.From && state.isCurrentStage(b.options.Target) {
-			break
-		}
+			fmt.Fprintf(b.Stdout, "%v / %v %v", currentCommandIndex, totalCommands, cmd)
```
"Step %d/%d : %v" for compatibility?

I think there are user applications that depend on the current logging format.

(should be converted to machine-readable format in the new RPC)

simonferquel replied:
There is actually a failing test for that that I am going to fix

simonferquel commented:
Fixed in last commit:
- Unknown instruction is a typed error
- MetaArgs processing in a function
- Shlex created once out of the meta-args loop
- \n in Build canceled message
- --from in COPY command is not expanded
- reverted to the old logic of build stage indexing
- parseLabel has a closer logic to parseEnv

simonferquel commented:
Last push:
Various integration tests fixes + golint

simonferquel commented:
All unit tests are now back (and due to typedCommands, it is easier to isolate parsing and dispatching tests).
Note, we might want to write more instructions parsing tests, or implement more end to end parse/dispatch tests like the ones in evaluator_tests

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -208,65 +141,70 @@ func (b *Builder) getImageMount(fromFlag *Flag) (*imageMount, error) {
+	state := d.state
+	state.beginStage(cmd.StageName, image)
+	d.builder.buildArgs.ResetAllowed()
+	if state.runConfig.OnBuild != nil && len(state.runConfig.OnBuild) > 0 {
```
I think we should keep the `OnBuild` handling as a separate function. It makes the individual steps of `dispatchFrom()` a lot more clear

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -208,65 +141,70 @@ func (b *Builder) getImageMount(fromFlag *Flag) (*imageMount, error) {
 	}
-
-	return stageName, nil
+	state.hasDispatchedFrom = true
```
This looks to be unused. Please remove

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -24,182 +24,117 @@ import (
+	}
+}
+func (d *stageBuilder) updateRunConfig() {
+	d.state.runConfig.Image = d.state.imageID
```
I think this was more appropriate as a method on `dispatchState`

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -225,6 +160,8 @@ func (s *dispatchState) beginStage(stageName string, image builder.Image) {
 	s.baseImage = image
 	s.setDefaultPath()
+	s.runConfig.OpenStdin = false
+	s.runConfig.StdinOnce = false
```
Ah, nice. I hadn't realized these were previously mixed up in the onBuild logic. This is a much better place for them.

dnephin commented on builder/dockerfile/evaluator_test.go:
```diff
@@ -177,34 +175,27 @@ func executeTestCase(t *testing.T, testCase dispatchTestCase) {
-		buildArgs: newBuildArgs(options.BuildArgs),
+	if err != nil {
+		if !strings.Contains(err.Error(), testCase.expectedError) {
+			t.Fatalf("Wrong error message. Should be \"%s\". Got \"%s\"", testCase.expectedError, err.Error())
```
```go
testutil.ErrorContains(t, err, testCase.expectedError)
```

(and below as well)

dnephin commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,381 @@
+type KeyValuePairs []KeyValuePair
+
+// BaseCommand is the base class for Dockerfile (String() returns its source code)
+type BaseCommand struct {
```
nit: `Base` is not a descriptive name. Maybe just `named` since it's an instruction in the instructions package, and it's not used externally.

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -651,228 +408,120 @@ func healthcheck(req dispatchRequest) error {
-	if len(req.args) == 0 {
-		return errAtLeastOneArgument("EXPOSE")
+func (d *stageBuilder) dispatchExpose(c *instructions.ExposeCommand, envs []string) error {
+	// custom multi word expansion
```
Some text about why this is a special case and doesn't fit the expander interface would be great.

dnephin commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,381 @@
+type LabelCommand struct {
+	BaseCommand
+	Labels KeyValuePairs // kvp slice instead of map to preserve ordering
+}
```
`LabelCommand` and `EnvCommand` are identical (except for the field name, which could be the same).

We already have `name` to differentiate them, so instead of two identical structs could we make this a single struct. In dispatch we can use the name to dispatch to two different functions.

simonferquel replied:
I don't agree: the whole point of these PR is to have typed commands. Field names are part of this strongly typed approach, and giving fields generic names just to share the same structure for 2 different aliases seems a bit wrong to me. The extreme case would be to have a common `type Command2TextFields{ Field1, Field2 string}` and having some aliases to this type. But in that case, what is the point of having strongly typed commands if field names have no real meaning ? 

I prefer a bit more code whith more expressiveness and this coherent 1<=> 1 relashionship between typed commands and dockerfile instructions. In longer term it seems more readable / maintainable (what if we want to add a flag on ENV for exemple that has no meaning at all on Label at some point in the future ?)

dnephin replied:
> I don't agree: the whole point of these PR is to have typed commands.

We can still have strict types without the code duplication. As discussed in slack I think we can do this with embedded structs. The field name can just be `Value` and it doesn't weaken the type guarantees at all.


> what if we want to add a flag on ENV for exemple that has no meaning at all on Label at some point in the future ?

Then at that time we can create a new struct. There's no reason to maintain extra code in the mean time for a case that may never occur.

dnephin commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,381 @@
+	Srcs []string
+	Dest string
+	From string
+}
```
Same thing here with `Copy` and `Add`. We could remove `AddCommand` and use just a single struct dispatched by name.

dnephin commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,381 @@
+type WorkdirCommand struct {
+	BaseCommand
+	Path string
+}
```
All of these "commands" that have a single `string` (maintainer, onbuild, workdir, stop signal, user) field could be reduce to two structs. One that supports expand and one that doesn't. One could even compose the other.

dnephin commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,381 @@
+	BaseCommand
+	Cmd          strslice.StrSlice
+	PrependShell bool
+}
```
Cmd/RunEntrypoint are all the same struct , same idea here.

dnephin commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,652 @@
+	}, nil
+
+	//TODO
+	//req.buildArgs.ResetAllowed()
```
commented out code to remove

dnephin commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,652 @@
+
+}
+
+func parseCmd(req parseRequest) (*CmdCommand, error) {
```
Duplicate of `parseRun()` which would go away with the suggested struct changes for instructions.

simonferquel replied:
I want to keep multiple functions (for maintainability sake), but I will try to share some internal code here

dnephin replied:
How does more code make it easier to maintain?

simonferquel replied:
My point is if we add flags to `cmd` in the future that are not in `run`, it would be easier if the code is already split (even if in its implementation it calls a single shared function with `run` and `entrypoint`). It is also more explicit: if I'm looking to change `run` parsing behavior, I know that I can modify parseRun without any side effect on other command parsing logic

dnephin commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,652 @@
+	var value *string
+	if hasDefault {
+		value = &newValue
+	}
```
I think this if is unnecessary. Why not just set it in the branch above?

dnephin commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,652 @@
+}
+
+// IsUnknownInstruction checks if the error is an UnknownInstruction or a parseError containing an UnknownInstruction
+func IsUnknownInstruction(err error) bool {
```
Sorry, I know it was my suggestion to make this a type. I should have looked at the code some more. I think exporting the metrics function and calling it directly where the error is raised would be cleaner.

dnephin reviewed (commented):
This is looking good.

I think most of my comments are just about removing some duplication in the new instruction types.

If we really want to keep unique types for each. I think we could still do that with `type foo bar` aliases, instead of repeating the structs and the methods.

simonferquel commented on builder/dockerfile/builder.go:
```diff
@@ -168,62 +177,78 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
+		return err
+	}
+	b.buildArgs.AddArg(meta.Name, meta.Value)
+	b.buildArgs.AddMetaArg(meta.Name, meta.Value)
```
Hmm I like the idea (and it is required for build parallelisation). We still need a common buildArgs instance for metaArgs construction, but once it is done, we should simply clone it once per stage, and have individual stage have their own instance.

simonferquel commented on builder/dockerfile/builder.go:
```diff
@@ -269,39 +294,36 @@ func BuildFromConfig(config *container.Config, changes []string) (*container.Con
+		if err != nil {
+			if instructions.IsUnknownInstruction(err) {
+				buildsFailed.WithValues(metricsUnknownInstructionError).Inc()
+			}
```
Hmm I do not agree (and that is actually the opposite of @tonistiigi feedback on the same subject before). I'd like to keep the node -> typed command parsing logic the least dependent on anything else as possible.

tonistiigi replied:
It might have been a mistake to log metrics for `BuildFromConfig` as it is just wrongly named function that doesn't have anything to do with `docker build`. There should be only one place then. Also, maybe there is no need to check for the error type at all as the point should be to log metrics for all errors. The selection of errors that are logged separately and what are not is completely arbitrary anyway.

simonferquel replied:
the metrics name is explitily "unknown_instruction_error" should we add a "general_parsing_error" metric then for other parsing errors?

simonferquel commented on builder/dockerfile/dispatchers.go:
```diff
@@ -34,32 +31,14 @@ import (
-	}
-
-	runConfig := req.state.runConfig
+func (d *stageBuilder) dispatchEnv(c *instructions.EnvCommand) error {
```
I come from a very object oriented background where we tend to group state and methods together in a coherent stuff and avoid free functions, so I am not sure I 100% agree, but for consistency sake, I'll do that.

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -208,65 +141,76 @@ func (b *Builder) getImageMount(fromFlag *Flag) (*imageMount, error) {
-		return nil
+	state := d.state
+	state.beginStage(cmd.StageName, image)
+	if state.runConfig.OnBuild != nil && len(state.runConfig.OnBuild) > 0 {
```
This first nil check is unnecessary, you can just check length (https://play.golang.org/p/E0QPccyYAU)

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -208,65 +141,76 @@ func (b *Builder) getImageMount(fromFlag *Flag) (*imageMount, error) {
+		err = dispatchTriggeredOnBuild(d, triggers)
+		if err != nil {
+			return err
+		}
```
This `err` condition is unnecessary. You can just 

```go
return dispatchTriggeredOnBuild(d, triggers)
```

dnephin commented on builder/dockerfile/evaluator_test.go:
```diff
@@ -177,34 +176,23 @@ func executeTestCase(t *testing.T, testCase dispatchTestCase) {
-		Stdout:    ioutil.Discard,
-		buildArgs: newBuildArgs(options.BuildArgs),
+	if err != nil {
+		testutil.ErrorContains(t, err, testCase.expectedError)
```
I think, instead of the comments in the test cases above, it would be good to split this into two test cases. One for parse errors, and another for dispatch errors.

simonferquel replied:
Aggreed. I should move parsing error related tests to the instructions package as well, and change evaluator_tests to not rely on parsing, but on already constructed typed errors.

dnephin commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,381 @@
+type KeyValuePairs []KeyValuePair
+
+// WithNameAndCode is the base class for Dockerfile (String() returns its source code)
+type WithNameAndCode struct {
```
It seems this doesn't need to be exported.

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -192,12 +111,30 @@ type dispatchState struct {
+	shlex   *ShellLex
+	builder *Builder
+	source  builder.Source
+}
```
Could this be renamed back to `dispatchRequest` (and with the variable name in the dispatch functions back to `req`). I think that will help focus the diff on the changes related to the new typed instructions.

simonferquel replied:
Ok, I found the "stageBuild" name more explicit (it represents the whole state and dependencies required to build a stage), but I agree it would simplify the diff. Might be worth adding a Naming PR after that though

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -192,12 +111,30 @@ type dispatchState struct {
+	return &dispatchState{runConfig: &container.Config{}, buildArgs: args}
 }
 
+func newStageBuild(builder *Builder, escapeToken rune, source builder.Source, buildArgs *buildArgs) *stageBuild {
```
Minor nit. The definition for `dispatchState` and `StageBuild` are a bit tangled here. Could you move this above `newDispatchState()`, so their "constructors" are together with the struct.

tonistiigi commented on builder/dockerfile/evaluator.go:
```diff
@@ -25,178 +25,182 @@ import (
+		shlex:           NewShellLex(escapeToken),
+		builder:         builder,
+		source:          source,
+		previousResults: previousResults,
```
`stages` ?

tonistiigi commented on builder/dockerfile/buildargs.go:
```diff
@@ -42,6 +42,26 @@ func newBuildArgs(argsFromOptions map[string]*string) *buildArgs {
+	return result
+}
+
+func (b *buildArgs) MergeReferencedArgs(other *buildArgs) {
```
`MarkArgsAsUsed`

simonferquel replied:
Hmm, I find that more confusing than `MergeReferecedArgs`. The idea there is to merge the referenced args lists (each build stage has its own buildArgs instance now, so at the end, we merge the referenced args altogether to be able to warn if a given build arg was not consumed)

tonistiigi replied:
Ok, hopefully, this will go away with preprocessing the args in the future.

tonistiigi commented on builder/dockerfile/builder.go:
```diff
@@ -271,62 +274,97 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
 	}
 	return aux.Emit(types.BuildResult{ID: state.imageID})
 }
+func convertMapToEnvs(m map[string]string) []string {
```
maybe `convertMapToEnvList` ? as map and env don't seem convertable types

tonistiigi commented on builder/dockerfile/builder.go:
```diff
@@ -271,62 +274,97 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
+	return result
+}
+func processMetaArg(meta instructions.ArgCommand, shlex *ShellLex, args *buildArgs) error {
+	envs := convertMapToEnvs(args.GetAllAllowed())
```
Add comment that this is because `shlex` currently expects concatenated string format.

tonistiigi reviewed (commented):
Didn't manage to go over everything yet but generally looks good. Moving to code-review

simonferquel commented on builder/dockerfile/copy.go:
```diff
@@ -68,10 +68,10 @@ type copier struct {
 }
 
-func copierFromDispatchRequest(req dispatchRequest, download sourceDownloader, imageSource *imageMount) copier {
+func newCopier(source builder.Source, cache pathCache, download sourceDownloader, imageSource *imageMount) copier {
```
This is because it does not create a copier from a dispatchRequest anymore, but 2 components to reduce the necessary interface (although, it was initially done because at some point there was no dispatchRequest type anymore)

tonistiigi replied:
Seems to me that `newCopier` is still called from `dispatchRequest`. Just these properties are extracted on caller side.

simonferquel commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -0,0 +1,619 @@
+			}
+		}
+		switch c := cmd.(type) {
+		case *BuildableStage:
```
I like the fact that I have a single point of AST analysis. Is it ok if I rename ParseCommand to ParseStatement ? (to make it more less confusing when the parsed line is the beginning of a stage)

simonferquel commented on builder/dockerfile/evaluator_unix.go:
```diff
@@ -4,6 +4,6 @@ package dockerfile
 // platformSupports is a short-term function to give users a quality error
 // message if a Dockerfile uses a command not supported on the platform.
-func platformSupports(command string) error {
+func platformSupports(command interface{}) error {
```
Although it seems more logical to have platform support checks at this package level (the instructions package is only there for parsing and typed representation. it is the dispatcher itself that should know which command is supported or not on the target platform)

I don't think we can implement interfaces for a type living in a different package, can we ?

tonistiigi replied:
I'm not sure we want to let implementation decide what parts of the `Dockerfile` they implement. All implementations should have the same behavior for user commands.

simonferquel replied:
Yes and actually, with lcow support, we should change this behavior to be "target platform" dependent instead of statically defined

simonferquel replied:
Yes and actually, with lcow support, we should change this behavior to be "target platform" dependent instead of statically defined

simonferquel commented on builder/dockerfile/builder.go:
```diff
@@ -240,24 +240,27 @@ func (b *Builder) build(source builder.Source, dockerfile *parser.Result) (*buil
+		found, targetIx := instructions.HasStage(stages, b.options.Target)
+		if !found {
+			buildsFailed.WithValues(metricsBuildTargetNotReachableError).Inc()
+			return nil, errors.Errorf("failed to reach build target %s in Dockerfile", b.options.Target)
```
Can't do that, it is an existing error and there is a non-regression test looking for "failed to reach build target" in stderr

dnephin commented on builder/dockerfile/builder.go:
```diff
@@ -240,24 +240,27 @@ func (b *Builder) build(source builder.Source, dockerfile *parser.Result) (*buil
 	if err != nil {
+		if instructions.IsUnknownInstruction(err) {
+			buildsFailed.WithValues(metricsUnknownInstructionError).Inc()
+		}
```
There was some discussion of changing this to be a more generic "parse failed" error so that we don't need this condition. What was the outcome of that?

simonferquel replied:
Don't know. I am not sure if it is a good idea to change existing metrics though. Adding a generic parsing failed metric could make sense without removing this one I suppose.

dnephin commented on builder/dockerfile/builder.go:
```diff
@@ -272,61 +275,91 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
+
+			fmt.Fprintf(b.Stdout, stepFormat, currentCommandIndex, totalCommands, cmd)
+			currentCommandIndex++
+			fmt.Fprintln(b.Stdout)
```
These 3 lines are occur 3 times in this function. Should be split out into a new function.

dnephin commented on builder/dockerfile/builder.go:
```diff
@@ -383,39 +416,33 @@ func BuildFromConfig(config *container.Config, changes []string) (*container.Con
 
-	if err := checkDispatchDockerfile(dockerfile.AST); err != nil {
-		return nil, err
+	commands := []interface{}{}
```
Don't we have an interface here instead of `interface{}` ?

simonferquel replied:
now that every command embeds withNameAndCode, I suppose I can introduce an interface yes

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -25,178 +25,182 @@ import (
-	imageID    string
-	baseImage  builder.Image
-	stageName  string
+func newDispatchRequest(builder *Builder, escapeToken rune, source builder.Source, buildArgs *buildArgs, stages *stagesBuildResults) *dispatchRequest {
```
Does this need to be a pointer? Previously it was immutable.

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -25,178 +25,182 @@ import (
-func newDispatchState() *dispatchState {
-	return &dispatchState{runConfig: &container.Config{}}
+func (d *dispatchRequest) commitStage() error {
+	return d.stages.commitStage(d.state.stageName, d.state.runConfig)
```
`commitStage()` sounds like a very strange behaviour for a `dispatchRequest()`. This probably belongs somewhere else.

dnephin commented on builder/dockerfile/evaluator.go:
```diff
@@ -25,178 +25,182 @@ import (
-		shlex:      options.shlex,
-		state:      options.state,
-		source:     options.source,
+func dispatch(d *dispatchRequest, cmd interface{}) error {
```
`interface{}` could be an interface that is satisfied by `withNameAndCode`, right?

dnephin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -548,6 +548,64 @@ RUN [ $(ls -l /exists/exists_file | awk '{print $3":"$4}') = 'dockerio:dockerio'
 		build.WithFile("test_file4", "test4")))
 }
 
+func (s *DockerSuite) TestBuildCopyCacheOnFileChange(c *check.C) {
```
All new integration tests should go into `docker_api_build...` and use the API instead of the CLI

dnephin commented:
What is the distinct between a command and an instruction?

simonferquel commented:
@dnephin: an instruction is any valid statement in a dockerfile including stage begin markers (`from` instructions)
Commands are instructions that have a side effect to the stage they belong to (all instructions except `from`)

So, `from` is an instruction but not a command -> it is translated to `instructions.Stage`
Other instructions are commands -> they are translated to `instructions.Command` and belong to an `instructions.Stage`

simonferquel commented:
@tonistiigi, @dnephin have you any more feedback or should we make this PR advance one more step toward merging ?

shouze commented:
Yup, would be nice to see this one merged

AkihiroSuda commented:
@tonistiigi Is this ready to merge?

simonferquel commented:
Yet another rebase for the chown flag on add/copy. Let us see how it behaves in CI

tonistiigi commented on builder/dockerfile/evaluator.go:
```diff
@@ -20,183 +20,180 @@
 	"strings"
 
+	"reflect"
+
```
nit: extra whitespace

tonistiigi commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -0,0 +1,394 @@
+
+// Sources list the source paths
+func (s SourcesAndDest) Sources() []string {
+	return s[:len(s)-1]
```
This should make a copy to avoid caller doing an append leaking to destination variable.

tonistiigi reviewed (approved):
LGTM

Follow ups could be in areas of moving more argument expansion to the `instructions` pkg and generalizing the build args. These seem areas where a different dispatcher still needs to duplicate some logic.

ping @dnephin

yongtang commented:
@simonferquel Can you squash the PR commits? The lint/rebase fixes might be too verbose to show up in the git history.

AkihiroSuda reviewed (approved):
LGTM
